src/extractor/guidance/coordinate_extractor.cpp

```diff
@@ -0,0 +1,834 @@
+    const auto regression_line = RegressionLine(coordinates);
+
+    /*
+     * If we can find a line that represents the full set of coordinates within a certain range in
+     * relation to ASSUMED_LANE_WIDTH, we use the regression line to express the turn angle.
+     * This yields a transformation similar to:
+     *
+     *         c   d                       d
+     *    b              ->             c
+     *                               b
+     * a                          a
+     */
+    const double max_deviation_from_regression = GetMaxDeviation(
+        coordinates.begin(), coordinates.end(), regression_line.first, regression_line.second);
+
+    if (max_deviation_from_regression < 0.35 * ASSUMED_LANE_WIDTH)
+    {
+        // We use the locations on the regression line to offset the regression line onto the
+        // intersection.
+        const auto coord_between_front =
+            util::coordinate_calculation::projectPointOnSegment(
+                regression_line.first, regression_line.second, coordinates.front())
+                .second;
+        const auto coord_between_back =
+            util::coordinate_calculation::projectPointOnSegment(
+                regression_line.first, regression_line.second, coordinates.back())
+                .second;
+        return GetCorrectedCoordinate(turn_coordinate, coord_between_front, coord_between_back);
+    }
+
+    const auto total_distance =
+        std::accumulate(segment_distances.begin(), segment_distances.end(), 0.);
+
+    if (IsDirectOffset(coordinates,
+                       straight_index,
+                       straight_distance,
+                       total_distance,
+                       segment_distances,
+                       considered_lanes))
+    {
+        // could be too agressive? Depend on lanes to check how far we want to go out?
+        // compare
+        // http://www.openstreetmap.org/search?query=52.411243%2013.363575#map=19/52.41124/13.36357
+        const auto offset_index = std::max<decltype(straight_index)>(1, straight_index);
+        return GetCorrectedCoordinate(
+            turn_coordinate, coordinates[offset_index], coordinates[offset_index + 1]);
+    }
+
+    if (IsCurve(coordinates,
+                segment_distances,
+                total_distance,
+                considered_lanes * 0.5 * ASSUMED_LANE_WIDTH,
+                turn_edge_data))
+    {
+        /*
+         * In curves we now have to distinguish between larger curves and tiny curves modelling the
+         * actual turn in the beginnig.
+         *
+         * We distinguish between turns that simply model the initial way of getting onto the
+         * destination lanes and the ones that performa a larger turn.
+         */
+        const double offset = 0.5 * considered_lanes * ASSUMED_LANE_WIDTH;
+        coordinates = TrimCoordinatesToLength(std::move(coordinates), offset);
+        const auto vector_head = coordinates.back();
```

Comment:
From chat: `.back()` is valid no matter what. Added an assert in TrimCoordinates to ensure we still have an element.